skrifa/src/color/mod.rs

```diff
@@ -218,7 +237,27 @@ pub trait ColorPainter {
     fn pop_clip(&mut self);
 
     /// Fill the current clip area with the specified gradient fill.
-    fn fill(&mut self, brush: Brush);
+    fn fill(&mut self, brush: Brush<'_>);
+
+    /// Queries from the client whether the optimized
+    /// [`fill_glyph()`](ColorPainter::fill_glyph) operation is supported.
+    fn fill_glyph_supported(&self) -> FillGlyph {
+        FillGlyph::Unimplemented
+    }
+
+    /// Combined clip and fill operation.
+    ///
+    /// Apply the clip path determined by the specified `glyph_id`,
+    /// then fill it with the specified [`brush`](Brush), applying the `_transform`
+    /// transformation matrix to the brush.
+    fn fill_glyph(
```

Comment:
Since we can provide a default fallback for this using only methods on self, my preference would be to go that route here rather than adding additional methods, types and logic to query support. 

Something like:
```rust
self.push_clip_glyph(glyph_id);
// maybe check for identity?
self.push_transform(transform);
self.fill(brush);
self.pop_transform();
self.pop_clip();
```

This allows the user to override as desired and removes the need for the result return type and the additional error variant.

Reply:
I thought about that, and also had this implemented first with just a client callback that can return a failure if it doesn't support the optimization, which then lead to the call sequence you're proposing, so that was semantically similar, but I see a case where this leads to less efficient outcomes.

If the client does not support the operation, we spend extra time for the detection while there is no optimization benefit - so a separate detection query saves time here. 

This also practically happens in the real world: In the no-clipbox situation of the Skia-side `BoundsPainter` where the draw method is used with no-op gradient fills. Here, the detection of the optimizable leaf situation only costs overhead, hence I thought it's better to ask first and avoid the search.

Reply:
> If the client does not support the operation, we spend extra time for the detection while there is no optimization benefit - so a separate detection query saves time here.

If a client doesn't support the optimization then it's probably not particularly concerned about performance? :)

> This also practically happens in the real world: In the no-clipbox situation of the Skia-side BoundsPainter where the draw method is used with no-op gradient fills. Here, the detection of the optimizable leaf situation only costs overhead, hence I thought it's better to ask first and avoid the search.

If we're worried about performance in the bounds computation case, I'd be more concerned about the color stop evaluation which I suspect is significantly more expensive than leaf detection. This might suggest that we want a traversal designed specifically for bounding boxes that avoids unnecessary work.

Overall, this feels like a case where the cost of using a nested traversal is leaking into the public API and that makes me a bit uneasy. My preference would be to do the idiomatic Rust thing here (default trait method for fallback) and then tune the implementation (which may imply specialized bounds computation) if performance is an issue.

Reply:
Updated according to yours and Ben's feedback. Providing a default implementation and assuming in traversal that the method is supported.